// travel-log/defines/constants.js
/*********Used in email validation, pwd mgmt  API Begin ************/
exports.DB_USER_SUCCESSFULLY_VALIDATED            =  1;
exports.DB_USER_ALREADY_VALIDATED                 =  2;
exports.DB_USER_ACCOUNT_NOT_VERIFIED              = -2;
exports.DB_USER_ACCOUNT_DOES_NOT_EXISTS           = -1;
exports.DB_USER_ACCOUNT_NOT_VERIFIED              = -2;
exports.DB_INVALID_USER_IN_LINK_PROVIDED          = -1;
exports.DB_INVALID_UUID_IN_LINK_PROVIDED          = -2;
exports.DB_EXPIRED_UUID_IN_LINK_PROVIDED          = -3;
exports.DB_NUM_AFFECTED_ROWS_IS_ONE               =  1;
exports.DB_USER_LINK_SUCCESSFULLY_VALIDATED       =  1;
exports.DB_FORGOT_PWD_INVALID_LINK                = -1;
exports.DB_PWD_UPDATE_ACTION_SUCCESS              =  1;
/**********Used in email validation , pwd mgmt API End**************/


/********** Used in user signup namma-auth controller Begin ********/

//Everything is a success and a new user entry is created
exports.DB_NEW_USER_ADDED_SUCCESSFULLY            =  1;

//An user with this email id already exists
exports.DB_USER_EMAIL_ALREADY_EXISTS              = -1;

/********* Used in user signup namma-auth controller End ***********/


/********* Used in user Login namma-auth controller Begin **********/
//Account exits and user email is validated
exports.DB_USER_ACCOUNT_EXISTS_AND_IS_VALIDATED   =  1;

//No record in user list for the given username
exports.DB_USER_ACCOUNT_DOES_NOT_EXISTS           = -1;

//User email has not been validated
exports.DB_USER_ACCOUNT_NOT_VERIFIED              = -2;

//User email has not been validated
//Also, the link sent to user has expired
exports.DB_USER_ACCOUNT_NOT_VERIFIED_LINK_EXPIRED = -3;

//User type provided is wrong, user exists with
// different user type ex: fb, goog etc
exports.DB_USER_EXISTS_FOR_DIFFERENT_USER_TYPE    = -4;

/*********** Used in user Login namma-auth controller End **********/


/* Session Defines Begin */
exports.SESSION_USER_NOT_VALIDATED    = 0;
exports.SESSION_USER_VALIDATED        = 1;
exports.SESSION_NAMMA_USER            = 1;
exports.SESSION_USER_LOGIN            = 1;
exports.SESSION_USER_LOGOUT           = 0;
/*Session Defines End*/

/* User Type Define Begin */

exports.USER_TYPE = {
  'regular' : 1,
  'facebook': 2,
  'google+' : 3,
  'twitter' : 4
};
/* User Type Define End */

/* User Gender Define Begin  */

exports.GENDER = {
  'male'  : 1,
  'female': 2,
  'none'  : 3  
};

/* User Gender Define End */

exports.SP_OPCODE = {
  'add'   : 1,
  'delete': 2,
  'get'   : 3,
  'update': 4    
}

/* Temporary Device ID */
exports.SP_USER_SIGN_UP_TEMP_DEVICE_ID = 1;
exports.SOCIAL_SIGNUP_USER_TEMP_IMEI_NUMBER = 123456789;
exports.SP_DB_STATUS_INVALID_IMEI = -1;
exports.SP_DB_STATUS_DEVICE_ID_USER_ID_MISMATCH = -2;

/* User Auth Response Status */
exports.SP_USER_AUTH_STATUS_SUCCESS                                 =  1;
exports.SP_USER_AUTH_INVALID_PASSWORD                               = 'invalid';
exports.SP_USER_AUTH_INVALID_USER_ID                                = -1;
exports.SP_USER_AUTH_STATUS_ACCOUNT_EXISTS                          = -1;
exports.SP_USER_AUTH_STATUS_ACCOUNT_NOT_VERIFIED                    = -2;
exports.SP_USER_AUTH_STATUS_ACCOUNT_NOT_VERIFIED_LINK_EXPIRED       = -3;
exports.SP_USER_AUTH_STATUS_ACCOUNT_EXISTS_FOR_DIFFERENT_USER_TYPE  = -4;

/* This tells if the Domain is Development(localhost) or Production(rhcloud) */
exports.LOCALHOST           =  'localhost';
exports.PRODUCTION          =  'parkking-lotr.rhcloud.com' //'www.scubenow.com'; ;
exports.LOCALHOST_URL       =  'http://'+exports.LOCALHOST+':3000';
exports.PRODUCTION_URL      =  'http://'+exports.PRODUCTION;
exports.DOMAIN;             
exports.DOMAIN_URL;
exports.FACEBOOK_PROFILE_PIC_URL = "https://graph.facebook.com/$/picture?type=large"; //Replace $ with Id
exports.PRODUCTION_ALIAS    =  'www.scubenow.com';


// Json types for json builder util
exports.JSON_TYPE = {
  'singleArraySingleObj'   : 1,
  'singleArrayMultipleObj' : 2,
  'multiArraySingleObj'    : 3,
  'multiArrayMultipleObj'  : 4    
}

// Blogs
exports.APIS_BLOGS_QUERY_CASE_BY_VIEW_COUNT          = '1';
exports.APIS_BLOGS_QUERY_CASE_BY_LOCATION            = '2';
exports.APIS_BLOGS_QUERY_CASE_BY_TOP_LOCATION        = '3';
exports.APIS_BLOGS_QUERY_CASE_BY_USER_NAME           = '4';
exports.APIS_BLOGS_QUERY_CASE_DISPLAY_GET_FORM       = '5'; // To view others blogs
exports.APIS_BLOGS_QUERY_CASE_BY_PROXIMITY           = '6'; 

exports.APIS_FOOD_QUERY_CASE_BY_VISIT_COUNT          = '1';
exports.APIS_FOOD_QUERY_CASE_BY_LOCATION             = '2';

exports.APIS_BLOGS_POST_ACTION_TYPE_INSERT           = '1';
exports.APIS_BLOGS_POST_ACTION_TYPE_DELETE           = '2';
exports.APIS_BLOGS_POST_ACTION_TYPE_UPDATE           = '3';

exports.APIS_POSTS_LIKES_INCREMENT_FOR_TRAVEL_PLACE  = '1';
exports.APIS_POSTS_LIKES_INCREMENT_FOR_FOOD_PLACE    = '2';


// User Profile
exports.APIS_USER_PROFILE_GET_QUERY_CASE_BY_USER_ID = '1';
exports.APIS_USER_PROFILE_GET_QUERY_CASE_BY_QB_ID   = '2';
exports.APIS_USER_PRIVACY_GET_QUERY_CASE_BY_USER_ID = '1';
exports.APIS_USER_FAVORITES_CATEGORY_GET_BY_USER_ID = '1';
exports.APIS_USER_FAVORITES_CATEGORY_GET_ALL        = '2';

exports.APIS_USER_FAVORITES_POST_ACTION_TYPE_INSERT = '1';
exports.APIS_USER_FAVORITES_POST_ACTION_TYPE_DELETE = '2';
exports.APIS_USER_FAVORITES_POST_ACTION_TYPE_GET    = '3';
