// travel-log/defines/dir.js
/*********************Global File for all Consumer APP Scube directory paths*************************/
///////////////////////////////////////////////////////////////////////////////////////////

/*******************************Scube APIS Paths START**************************************/
exports.API_HELPER_LOCATIONS                   =  'travel-log/api-helpers/locations/locations-helper';
exports.API_HELPER_LOGIN_DEVICE_MANAGEMENT     =  'travel-log/api-helpers/device/device-mgmt-helper';
exports.API_HELPER_MALLS                       =  'travel-log/api-helpers/malls/malls-helper';
exports.API_HELPER_SHOPS                       =  'travel-log/api-helpers/shops/shops-helper';
exports.API_HELPER_BRANDS                      =  'travel-log/api-helpers/brands/brands-helper';
exports.API_HELPER_SCUBITS_GET                 =  'travel-log/api-helpers/scubits/scubits-get-helper';
exports.API_HELPER_SCUBITS_CONTEXT_GET         =  'travel-log/api-helpers/scubits/scubits-context-get-helper';
exports.API_HELPER_SCUBITS_POST                =  'travel-log/api-helpers/scubits/scubits-post-helper';
exports.API_HELPER_SEARCH                      =  'travel-log/api-helpers/search/search-helper';
exports.API_HELPER_USER_PROFILE_GET            =  'travel-log/api-helpers/user/profile/user-profile-get-helper';
exports.API_HELPER_USER_PROFILE_UPDATE         =  'travel-log/api-helpers/user/profile/user-profile-update-helper';
exports.API_HELPER_IMAGES                      =  'travel-log/api-helpers/images/images-helper';
exports.API_HELPER_USER_PRIVACY_GET            =  'travel-log/api-helpers/user/privacy/user-privacy-get-helper';
exports.API_HELPER_USER_PRIVACY_UPDATE         =  'travel-log/api-helpers/user/privacy/user-privacy-update-helper';
exports.API_HELPER_USER_FAVORITES_GET          =  'travel-log/api-helpers/user/favorites/user-favorites-get-helper';
exports.API_HELPER_USER_FAVORITES_UPDATE       =  'travel-log/api-helpers/user/favorites/user-favorites-update-helper';
exports.API_HELPER_USER_ACCOUNT_GET            =  'travel-log/api-helpers/user/account/user-account-get-helper';
/*******************************Scube API's Paths END**************************************/

/*******************************Scube Controllers Paths START**************************************/
//C Implies its controller
//Login Controller
exports.C_LOGIN_NAMMA_AUTH              =  'travel-log/controllers/login/namma-auth-controller';
exports.C_LOGIN_SOCIAL_AUTH             =  'travel-log/controllers/login/social-auth-controller';
exports.C_LOGIN_PWD_MANAGEMENT          =  'travel-log/controllers/login/password-mgmt-controller';
exports.C_LOGIN_EMAIL_VALIDATION        =  'travel-log/controllers/login/email-validation-controller';
/*******************************Scube Controllers Paths END**************************************/


/*******************************Scube Library Paths BEGIN*************************************/
//Logger path
exports.LOGGER                          =  'travel-log/libs/logger/logger';
//Database Paths
exports.DB_SP                           =  'travel-log/libs/database/store-procedure';
exports.DB_CONNECT                      =  'travel-log/libs/database/mysql-db';
//Util Paths
exports.HASHER                          =  'travel-log/libs/utils/hasher';
exports.MAILER                          =  'travel-log/libs/utils/mailer';
exports.JSON_UTIL                       =  'travel-log/libs/utils/json';
exports.DOMAIN_HANDLER                  =  'travel-log/libs/utils/domain-handler';
exports.DATA_PARSER                     =  'travel-log/libs/utils/data-parser';
exports.JSON_BUILDER                    =  'travel-log/libs/utils/json-builder';
//GCM Paths
exports.GCM_POST                        =  'travel-log/libs/gcm/gcm-post';
/*******************************Scube Library Paths END***************************************/

/*******************************Scube Config Paths START**************************************/
exports.DB_CONFIG                       =  'travel-log/config/database-config';
exports.SOCIAL_CONFIG                   =  'travel-log/config/social-config';
exports.GCM_CONFIG                      =  'travel-log/config/gcm-config';
/*******************************Scube Library Paths END***************************************/

/*******************************Scube Services Paths START************************************/
exports.SERVICES_LOCATION               =  'travel-log/services/locations-service';
exports.SERVICES_DEVICE                 =  'travel-log/services/device-service';
exports.SERVICES_DEALS                  =  'travel-log/services/deals-service';
/*******************************Scube Routes Paths START************************************/
//R implies its routes

//Used in APP.js file
exports.R_LOGIN                         = './routes/login/'
exports.R_USER                          = './routes/user/'
exports.R_APIS                          = './routes/apis/'
exports.R_TESTS                         = './routes/tests/'
//APIS Routes
exports.R_API_LOCATIONS                 = './locations/locations-api-route';
exports.R_API_DEVICE_MGMT               = './device/device-mgmt-api-route';
exports.R_API_GCM                       = './gcm/gcm-api-route';
exports.R_API_MALLS                     = './malls/malls-api-route';
exports.R_API_SHOPS                     = './shops/shops-api-route';
exports.R_API_BRANDS                    = './brands/brands-api-route';
exports.R_API_SCUBITS_GET               = './scubits/scubits-get-api-route';
exports.R_API_SCUBITS_CONTEXT_GET       = './scubits/scubits-context-get-api-route';
exports.R_API_SCUBITS_POST              = './scubits/scubits-post-api-route';
exports.R_API_SEARCH                    = './search/search-api-route.js';
exports.R_API_IMAGES                    = './images/images-api-route.js';

//Login Action Routes
exports.R_LOGIN_NAMMA_AUTH_ACTION            =  './namma-auth-action-route';
exports.R_LOGIN_PASSWORD_MGMT_ACTION         =  './password-mgmt-action-route';
exports.R_LOGIN_EMAIL_VALIDATION_ACTION      =  './email-validation-action-route';

// Users Routers
exports.R_USER_PROFILE_GET              = './profile/user-profile-get-route';
exports.R_USER_PROFILE_UPDATE           = './profile/user-profile-update-route';
exports.R_USER_PRIVACY_GET              = './privacy/user-privacy-get-route';
exports.R_USER_PRIVACY_UPDATE           = './privacy/user-privacy-update-route';
exports.R_USER_FAVORITES_GET            = './favorites/user-favorites-get-route';
exports.R_USER_FAVORITES_UPDATE         = './favorites/user-favorites-update-route';
exports.R_USER_ACCOUNT_GET			        = './account/user-account-get-route';

//Tests Routes
exports.R_TESTS_DB                      =  './db';
exports.R_TESTS_REST_SERVICE_API        =  './rest-service-api';
exports.R_TESTS_GCM                     =  './testgcm';
exports.R_TESTS_DB_MONGO                =  './mongodb';

/*******************************Scube Routes Paths END**************************************/
