// travel-log/defines/dir.js
/*********************Global File for all Consumer APP Scube directory paths*************************/
///////////////////////////////////////////////////////////////////////////////////////////

/*******************************Scube APIS Paths START**************************************/
exports.API_HELPER_BLOGS_GET                   =  'travel-log/api-helpers/blogs/blogs-get-helper';
exports.API_HELPER_BLOGS_POST                  =  'travel-log/api-helpers/blogs/blogs-post-helper';
exports.API_HELPER_USER_PROFILE_GET            =  'travel-log/api-helpers/user/profile/user-profile-get-helper';
exports.API_HELPER_USER_PROFILE_UPDATE         =  'travel-log/api-helpers/user/profile/user-profile-update-helper';
exports.API_HELPER_USER_PRIVACY_GET            =  'travel-log/api-helpers/user/privacy/user-privacy-get-helper';
exports.API_HELPER_USER_PRIVACY_UPDATE         =  'travel-log/api-helpers/user/privacy/user-privacy-update-helper';
exports.API_HELPER_USER_FAVORITES_GET          =  'travel-log/api-helpers/user/favorites/user-favorites-get-helper';
exports.API_HELPER_USER_FAVORITES_UPDATE       =  'travel-log/api-helpers/user/favorites/user-favorites-update-helper';
exports.API_HELPER_USER_ACCOUNT_GET            =  'travel-log/api-helpers/user/account/user-account-get-helper';
/*******************************Scube API's Paths END**************************************/

/*******************************Scube Controllers Paths START**************************************/
//C Implies its controller
//Login Controller
exports.C_LOGIN_NAMMA_AUTH              =  'travel-log/controllers/login/namma-auth-controller';
exports.C_LOGIN_SOCIAL_AUTH             =  'travel-log/controllers/login/social-auth-controller';
exports.C_LOGIN_PWD_MANAGEMENT          =  'travel-log/controllers/login/password-mgmt-controller';
exports.C_LOGIN_EMAIL_VALIDATION        =  'travel-log/controllers/login/email-validation-controller';
/*******************************Scube Controllers Paths END**************************************/


/*******************************Scube Library Paths BEGIN*************************************/
//Logger path
exports.LOGGER                          =  'travel-log/libs/logger/logger';
//Database Paths
exports.DB_SP                           =  'travel-log/libs/database/store-procedure';
exports.DB_CONNECT                      =  'travel-log/libs/database/mysql-db';
//Util Paths
exports.HASHER                          =  'travel-log/libs/utils/hasher';
exports.MAILER                          =  'travel-log/libs/utils/mailer';
exports.JSON_UTIL                       =  'travel-log/libs/utils/json';
exports.DOMAIN_HANDLER                  =  'travel-log/libs/utils/domain-handler';
exports.DATA_PARSER                     =  'travel-log/libs/utils/data-parser';
exports.JSON_BUILDER                    =  'travel-log/libs/utils/json-builder';
//GCM Paths
exports.GCM_POST                        =  'travel-log/libs/gcm/gcm-post';
/*******************************Scube Library Paths END***************************************/

/*******************************Scube Config Paths START**************************************/
exports.DB_CONFIG                       =  'travel-log/config/database-config';
exports.SOCIAL_CONFIG                   =  'travel-log/config/social-config';
exports.GCM_CONFIG                      =  'travel-log/config/gcm-config';
/*******************************Scube Library Paths END***************************************/

//R implies its routes

//Used in APP.js file
exports.R_LOGIN                         = './routes/login/'
exports.R_USER                          = './routes/user/'
exports.R_APIS                          = './routes/apis/'

exports.R_API_BLOGS_GET                 = './blogs/blogs-get-api-route';
exports.R_API_BLOGS_POST                = './blogs/blogs-post-api-route';


//Login Action Routes
exports.R_LOGIN_NAMMA_AUTH_ACTION            =  './namma-auth-action-route';
exports.R_LOGIN_PASSWORD_MGMT_ACTION         =  './password-mgmt-action-route';
exports.R_LOGIN_EMAIL_VALIDATION_ACTION      =  './email-validation-action-route';

// Users Routers
exports.R_USER_PROFILE_GET              = './profile/user-profile-get-route';
exports.R_USER_PROFILE_UPDATE           = './profile/user-profile-update-route';
exports.R_USER_PRIVACY_GET              = './privacy/user-privacy-get-route';
exports.R_USER_PRIVACY_UPDATE           = './privacy/user-privacy-update-route';
exports.R_USER_FAVORITES_GET            = './favorites/user-favorites-get-route';
exports.R_USER_FAVORITES_UPDATE         = './favorites/user-favorites-update-route';
exports.R_USER_ACCOUNT_GET			        = './account/user-account-get-route';

/*******************************Scube Routes Paths END**************************************/
