// travel-log/libs/database/store-procedure.js
// Usage in your .js file
// var sp = require('scube-lib/store-procedure');
// var userLoginSp = new sp(<procedure-name>);
//
// Start adding parameters. second parameter determines if the first parameter should be passed
// as a string or a number to sp
// userLoginSp.add("value", true);
// userLoginSp.add(variable, false);
//
// Finally invoke 'call' to get the sp string
// var userLoginCall = userLoginSp.call();
//
// Make use of this function in connection.query
// connection.query(userLoginCall, function(err, row){
//  // your query response handling code
// });

var storeProcedure = function(name) {
  // Name of the store procedure
  this.spName = name;

  // Parameters
  this.params = [];

  // Single sp call lock
  this.invoked = false;

  // Add parameters to parameters array
  this.add = function(value, isStr) {
    // These logs are not needed now since calling sp log
    // below will have all the values
    //console.log("Adding "+value+" to sp = "+this.spName);
    
    // Convert parameter based on "isStr" flag
    if(isStr) {
      value = "'"+value+"'";
    }

    // Add value to array
    this.params.push(value);
  }

  // Add NULL's for given number of elements in parameters array
  this.addNull = function(num) {
    //console.log("Adding NULL to sp = "+this.spName);
    
    for(i=0;i<num;i++){
      //console.log("Pushing Null\n");
      // Add NULL to array
      this.params.push("NULL");
    }

  }

  // Add NULL's if the elemet is null 
  // Add value if element is not null
  this.addOrAddNull = function(element, isString) {
    //console.log("Adding NULL to sp = "+this.spName);
    if(!element){
      this.addNull(1);
    } else {
      this.add(element, isString);
    }
  }

  this.getSpParamsString = function() {
    // If input is not an array, convert it into an array for easy handling
    if(!(this.params instanceof Array)) {
      this.params = [this.params];
    }

    // Convert array into ',' separated string
    this.params = this.params.join();

    return this.params;
  }

  // Call Store procedure
  // Success : returns the store procedure call string
  // failure : returns false
  this.call = function() {
    // Flag : call should be invoked only once per "storeProcedure" instance
    if(!this.invoked)  this.invoked = true;

    // Convert params to ',' separated string
    var paramsStr = this.getSpParamsString();

    // empty string error handling
    if(typeof paramsStr === 'undefined' || typeof this.spName === 'undefined' || this.spName === '') {
      return false;
    }

    // Form the sp call according to syntax
    // Syntax : "CALL <sp-name>(params--comma-separated-string)"
    var spCallStr = "CALL "+this.spName+"("+paramsStr+")";

    console.log('SP invoked : '+spCallStr);

    return spCallStr; 
  }
};

module.exports = storeProcedure;
