// travel-log/libs/utils/mailer.js
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// This Library has two functions: 
// 1) sendUnicastMail. This takes in text to send and email id to send to as input and returns success or failure.
// 2) sendMultiCastMail. This takes in text to send and email ids of all users in a list as input and returns success or failure.
//
// Usage: 
// var mailer = require('scube-lib/mailer');
// var mail = new mailer();
// var result = mail.sendUnicastMail("Hi there sparkker", "Hi From Sparkk!", user@domain.com");
// var result = mail.sendMultiCastMail("Hi there sparkkers", "Hi From Sparkk!", "user1@domain1.com, user2@domain2.com, user3@domain3.com");
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
var mailer = function() {
  // This function sends out an email to a specified receiver
  this.sendUnicastMail = function(textToSend, subjectToSend, emailToSend) {
    var nodemailer     = require('nodemailer');
    var smtpTransport  = require('nodemailer-smtp-transport');

    var transporter = nodemailer.createTransport(smtpTransport({
      //host: "box907.bluehost.com",  // hostname
      //secureConnection: true,       // use SSL
      //port: 465,                    // port for secure SMTP
      host : "a2plcpnl0135.prod.iad2.secureserver.net",//"smtpout.secureserver.net",
      //secureConnection: true,
      secure: true,
      port : "465",
      auth: {
        //user: "noreplysparkk@gmail.com",
        //pass: "noreplysparkk!!!!"
        user: "donotreply@scubenow.com",
        pass: "donotreply2015"
      }
    }));
    console.log("Inside mailer ");
    transporter.sendMail({
      from:     "no-reply <donotreply@scubenow.com", // sender address
      to:       emailToSend,                      // comma separated list of receivers
      subject:  subjectToSend,                    // Subject line
      text:     textToSend                        // plaintext body
    }, function(error, response) {
      if(error){
        console.log(error);
      } else {
        console.log(response);
        console.log("Message sent: " + response.message);
      }
    });
  }

  //This function builds the text and subject for user email validation email
  this.sendUserEmailValidationLink = function(uuidToSend, emailToSend, userId){

    var textToSend    = "Hello Scubber!\n Please validate your email by clicking on the link:\n"+
                        DEFS.CONST.DOMAIN_URL+"/validateUserEmail?uuid=" + uuidToSend + "&a2119518=" + userId;
    var subjectToSend = "Validate your Scube account";
    var mailResult    = this.sendUnicastMail(textToSend, subjectToSend, emailToSend);
  }

  //This function builds the text and subject for reset password email
  this.sendResetPwdLink= function(uuidToSend, emailToSend, userId){

    var textToSend    = "Hello Scubber!\n We have received your password reset request.\n"+
                        "Reset your password by clicking on the link:\n"+ DEFS.CONST.DOMAIN_URL +
                        "/resetPwd?uuid=" + uuidToSend + "&a2119518=" + userId; 
    var subjectToSend = "Reset your password at Scube";
    var mailResult    = this.sendUnicastMail(textToSend, subjectToSend, emailToSend);
  }

};
module.exports = mailer;
