// travel-log/config/database-config.js
// Expose the database configuration to our app directly via module.exports

module.exports = {

    /*'openshift': {
    prd: {
      host     : process.env.OPENSHIFT_MYSQL_DB_HOST,
      username : 'scube_prd',
      password : 'thirst_to_app_store_nov@2015',
      port     : process.env.OPENSHIFT_MYSQL_DB_PORT,
      database : 'scube'
    },

    dev: {
      host     : '127.0.0.1',
      username : 'scube_dev',
      password : 'thirst_to_deploy_nov@2015',
      port     : 3306,
      database : 'scube_dev'
    }
  },*/

  'openshift_mongo': {
    /*prd: {
      host     : process.env.OPENSHIFT_MONGODB_DB_HOST,
      username : 'scube_admin',
      password : 'insvikrant+*-/2015',
      port     : process.env.OPENSHIFT_MONGODB_DB_PORT,
      database : 'scube_mongodb_dev',
      appname  : 'parkking'
    },

    dev: {
      host     : '127.0.0.1',
      username : 'scube_admin',
      password : 'insvikrant+*-/2015',
      port     : 27017,
      database : 'scube_mongodb_dev',
      appname  : 'parkking'
    },*/

    mongolab: {
      host     : 'ds051553.mongolab.com',
      username : 'root',
      password : 'abc123',
      port     :  51553,
      database : 'travelog',
      appname  : 'travelog'
    }

  }
};

  /*"database":"mongodb://root:abc123@ds051553.mongolab.com:51553/travelog",
  "port": process.env.PORT || 3000,
  "secretKey":"Radhika"*/
